webpack.config.ts

```diff
@@ -31,27 +36,72 @@ function readCanisterId() {
   return canisterId;
 }
 
-// This plugin emulates the behaviour of http.rs while using webpack dev server locally
-// so we don't need to proxy to the backend canister for the index.html file.
-// This overcomes some issues with Safari and the CSP headers set in http.rs.
-class InjectCanisterIdPlugin {
+// A plugin that replaces content in HTML files
+class HtmlReplacePlugin {
+  constructor(
+    /** The function to apply on the HTML */
+    private f: (html: string) => string,
+    /** If specified, only replace HTML in this file */
+    private name?: string
+  ) {}
+
   apply(compiler: webpack.Compiler) {
-    compiler.hooks.compilation.tap("InjectCanisterIdPlugin", (compilation) => {
+    compiler.hooks.compilation.tap(this.constructor.name, (compilation) => {
       HtmlWebpackPlugin.getHooks(compilation).beforeEmit.tapAsync(
-        "InjectCanisterIdPlugin",
+        this.constructor.name,
         (data, cb) => {
-          data.html = data.html.replace(
-            '<script id="setupJs"></script>',
-            `<script data-canister-id="${readCanisterId()}" id="setupJs"></script>`
-          );
-
+          if (this.name === undefined || this.name === data.outputName) {
+            data.html = this.f(data.html);
+          }
           cb(null, data);
         }
       );
     });
   }
 }
 
+// A plugin that creates an HTML page based on our index.html template
+const htmlPlugin = ({ filename }: { filename: string }): HtmlWebpackPlugin =>
+  new HtmlWebpackPlugin({
+    template: "src/frontend/assets/index.html",
+    // Don't inject the index.js in production, because the canister actually injects it (see http.rs for more details)
+    // When true, injects (a hot-reloading version of) the built javascript, which in turn inserts the CSS (through "style-loader").
+    // This value is read by the template; see index.html for more details.
+    inject: !isProduction,
+    filename,
+  });
+
+// A plugin that creates a static page by injecting a static TemplateResult into a page
+const staticPagePlugin = (
+  pageName: string,
+  pageContent: TemplateResult
+): HtmlReplacePlugin =>
+  new HtmlReplacePlugin((html) => {
+    const content = Array.from(render(pageContent)).reduce(
+      (acc, v) => acc + v,
+      ""
+    );
+    return html.replace(
+      '<main id="pageContent" class="l-wrap" aria-live="polite"></main>',
```

Comment:
if someone replaces the style class or the tag `main` one day, this replace won't work anymore. would it be more handy to replace a placeholder within the tag?

```
// config
return html.replace(
      '<!-- I_WILL_BE_REPLACED -->',
      
// effective html      
<main id="pageContent" class="l-wrap" aria-live="polite">
   <!-- I_WILL_BE_REPLACED -->
</main>
```

Reply:
That's a good point. I decided to go with this because in the rest of the app we explicitly use `#pageContent` as a container where we `render()` from, so I wanted to make sure this is consistent (since the prerendered version needs to be inserted at the same spot). I'll think a bit more.

Reply:
@peterpeterparker thinking about this more, I realize I have no idea what's best here. I'd suggest we go with this for now, and adapt as we go along (i.e. when/if more pages are added). Wdyt?

Reply:
I can live with it, I already accepted the PR, sure go on